### PR TITLE
[Sdk] Added options for disabling signals

### LINKED ISSUE #3455: OpenTelemetry .NET Auto-Instrumentation/SDK Side-by-Side Support
# Feature Request

**Is your feature request related to a problem?**

OpenTelemetry .NET Auto-Instrumentation **dynamically loads** OpenTelemetry .NET SDK and its instrumentation to user process and helps collect telemetry from an application without modifying the code. Issue happens when the application has a reference to OpenTelemetry SDK and user want to enable Auto-Instrumentation. In this case both user application and Auto-Instrumentation will have its own instance of `TracerProvider` or `MeterProvider` and could collect duplicate telemetries. 

**Describe the solution you'd like:**

Auto-Instrumentation uses [.NET startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md) to load the OpenTelemetry SDK and initialize `TracerProvider`/`MeterProvider` before the main application's entry point. SDK providing an option to disable the creation of the `TracerProvider`/`MeterProvider` for user-initiated providers should solve this issue. This could be implemented by adding a static bool flag in `TracerProvider`/ `MeterProvider` to stop creation of these providers. This flag can be private or public, auto-instrumentation has ability to set this through reflection. If customer use auto-instrumentation, this flag could be set by auto-instrumentation in startup and it will avoid creation of TracerProvider from user application. 

## Additional Context

https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/633

cc: @open-telemetry/dotnet-instrumentation-approvers

### PR DESCRIPTION
**Pull request by CodeBlanch**
This is an alternative design for #3583 which fixes #3455.

## Changes

We have a lot of environment variables sprinkled around the repo. There are issues like #2980 which are because we access envvars directly instead of through `IConfiguration`. This PR builds on top of #3533 to make it possible to use `IConfiguration` to access settings.

What I did here is introduce `SdkOptions` which contains flags for disabling the signals. `SdkOptions` can be bound to `IConfiguration` for users that want to drive it through JSON (or whatever). It will also look for `OTEL_SDK_DISABLED` key using `IConfiguration` which means users can specify the flag through envvars, command-line, json, whatever. @alanwest I would imagine things like attribute limits would also go here and piggyback on the `IConfiguration` stuff?

Users calling the "Sdk.Create*ProviderBuilder" methods directly (including .NET Framework) will just get an `IConfiguration` automatically created from envvars.

Essentially what I'm trying to do here is establish a better options pattern we can use throughout the repo (for exporters, instrumentation, etc.).

For auto-instrumentation case, I added a static registration method so it can disable the flags at runtime. Not the most elegant mechanism, but probably the easiest for consumption.

## Examples

* Auto-instrumentation disabling SDK builders

```csharp
// Note: Must be done AFTER auto-instrumentation has made its providers, before users have made theirs
SdkOptions.RegisterConfigureCallback(options =>
{
    options.TracingEnabled = false;
    options.MetricsEnabled = false;
    options.LoggingEnabled = false;
});
```

* Binding `SdkOptions` to some `IConfiguration` section

```csharp
builder.Services.Configure<SdkOptions>(builder.Configuration.GetSection("OpenTelemetry"));
```

```json
{
   "OpenTelemetry": {
       "TracingEnabled": false,
       "MetricsEnabled": false,
       "LoggingEnabled": false
   }
}
```

* Setting `OTEL_SDK_DISABLED` through `IConfiguration`

Lot's of ways to do this. Command-line, using environment variables, etc. This is how it would look through appSettings.json:

```json
{
   "OTEL_SDK_DISABLED": true
}
```

Final `IConfiguration` will be a composite of all the sources users have configured. `SdkOptions` will look for that key, basically.

## What to consider when comparing to the alternative design...

The issue I saw with #3583 is it is kind of a one-off thing. Auto-instrumentation would be tied to private bools via reflection. It would need to be changed if https://github.com/open-telemetry/opentelemetry-specification/pull/2679 gets merged. It also seemed like a useful feature for users to be able to toggle SDK on/off via config. I tried to enable all of these things while also setting the foundation for better options/settings going forward.

### CODE REVIEW

File: src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs [hunk 1]
```diff
@@ -264,7 +272,24 @@ protected TracerProvider Build()
 
             var serviceProvider = services.BuildServiceProvider();
 
-            return new TracerProviderSdk(serviceProvider, ownsServiceProvider: true);
+            var sdkOptions = serviceProvider.GetRequiredService<IOptions<SdkOptions>>().Value;
+            if (!sdkOptions.TracingEnabled)
+            {
+                serviceProvider.Dispose();
+
+                return new NoopTracerProvider();
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs:280]: @rajkumar-rangaraj Your design was returning the `TracerProvider.Default` for this case. I had that, but decided to add a no-op instance instead. The reason being: User's code is probably calling dispose/shutdown on the instance returned from build invocation. Seemed risky to return them something shared which they might kill accidentally?

File: src/OpenTelemetry/OpenTelemetry.csproj [hunk 2]
```diff
@@ -19,6 +19,7 @@
     <PackageReference Include="System.Reflection.Emit.Lightweight" Version="$(SystemReflectionEmitLightweightPkgVer)" Condition="'$(TargetFramework)' != 'net6.0'" />
     <PackageReference Include="Microsoft.Extensions.Logging" Version="$(MicrosoftExtensionsLoggingPkgVer)" />
     <PackageReference Include="Microsoft.Extensions.Logging.Configuration" Version="$(MicrosoftExtensionsLoggingConfigurationPkgVer)" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="$(MicrosoftExtensionsConfigurationEnvironmentVariablesPkgVer)" />
```
Comment (CodeBlanch) [inline src/OpenTelemetry/OpenTelemetry.csproj:22]: Don't love adding this dependency, but we do make heavy use of environment variables so it didn't seem totally unreasonable?

Comment (alanwest) [reply src/OpenTelemetry/OpenTelemetry.csproj:22]: I'm not personally super concerned, but I have heard concern from end users regarding dependency bloat... 

We could experiment with breaking out SDK configuration into it's own package. This is _kind of_ similar to what Java has with their auto-config module https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure-spi

**Comment by alanwest** [comment]
> @alanwest I would imagine things like attribute limits would also go here and piggyback on the IConfiguration stuff?

Yes! This is exactly what I had envisioned. I'd noted this in the description of #3376. The classes I put here https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Configuration was meant to be temporary. The class you introduce in this PR is effectively what I had in mind. I can move this stuff to unify things once this PR lands.

**Pull request closed without merging**